### ISSUE TITLE
Require Pelita players to be executable scripts?

### ISSUE THREAD

**Issue by Debilski** [ASPP/pelita#331]
I’d like to put this up for discussion:

Currently, users are required to write a Python module (say, `MyPlayer`) which contains a method `factory`. When a game is started, Pelita runs an external script `module_player.py` with the module and the socket address as arguments. Eg: `./pelita_path/module_player.py MyPlayer:factory tcp://127.0.0.0:52000`. The advantage for the user is that they are required to write a module and don’t need to deal with any command line handling themselves. The module then is also more readily testable. Also, we are relatively sure that the Pelita and the player use the same libraries.

The disadvantages on Pelita’s side are:
a) We need to figure out the location of `./pelita_path/module_player.py`. This is harder than it seems, as it will also need to work from within the tests.
b) We need to provide code to specify the Python version, if this differs from Pelita’s version of Python.

An alternative would be to require the following boilerplate:

```
#!/usr/bin/env python3
from pelita import cmdline

from . import MyPlayer
# MyPlayer would at least contain a `MyPlayer.factory`
if __name__ == '__main__':
    cmdline.run(MyPlayer)
```

which would be called like

```
./path_to_player/players_module.py factory tcp://127.0.0.0:52000
```

and the `cmdline` would do the rest.

Notable points:
a) If the interpreter line is messed up or does not work on the target machine, things will not work.
b) If pelitagame is run with a specific Python version, the user’s code will be unaffected by this choice.
c) @venthur will say that he proposed this ages ago (https://github.com/ASPP/pelita/issues/272)

**Comment by keszybz** [comment]
On Sat, Feb 20, 2016 at 09:58:15AM -0800, Rike-Benjamin Schuppner wrote:
> Notable points:
> a) If the interpreter line is messed up or does not work on the target machine, things will not work.
> b) If pelitagame is run with a specific Python version, the user’s code will be unaffected by this choice.
> c) @venthur will say that he proposed this ages ago (https://github.com/ASPP/pelita/issues/272

I don't see any particular downsides to the proposed change.
OTOH, I don't see the reason for the change either. In particular, I don't
think we care about running the client with different python versions... If we
were in the business of providing an environment to run the agents that change
would have clear benefits, but for us providing an environment to run the agents
is just a side effect of providing an environment to learn python. So keeping
things simple and consistent between server & clients is more important.
But probably simplicity is the most important. I'd pick the simpler solution;
whether it be the old one or the proposed one.

Zbyszek

**Comment by otizonaizit** [comment]
Well, the downsides of the proposed change are that all of sudden 
students are confronted with specifying sockets and ports on the command 
line, and I assume that the port number may even become problematic in 
case the mess up and run two clints simultaneously or 
something./pelita_path/module_player.py like that.

But, more importantly, I don't see the downsides of the current state of 
affairs. 

>The disadvantages on Pelita’s side are:
>a) We need to figure out the location of 
>`./pelita_path/module_player.py`. This is harder than it seems, as it 
>will also need to work from within the tests.

I don't see where the problem lies in this. We are doing it since the 
very beginning of the history of Pelita, so what exactly has become 
harder recently?

>b) We need to provide code to specify the Python version, if this 
>differs from Pelita’s version of Python.
I don't see this as a problem. I would even go as far as saying that we 
could finally drop py2 compatibility completely and move to a py3 only 
source code. Clients will be run with the same Python version as the 
server, full stop. So I don't see any need to specify Python versions…

@Debilski: is there any concrete problem or use case that you are trying 
to address here? I don't get what do we gain by making the pelita player 
modules runnable without using the pelitagame script. The tangent issue 
#272 has been cleanly solved, so I don't think that counts.

**Comment by Debilski** [comment]
Am 21.02.2016 um 20:01 schrieb Tiziano Zito <notifications@github.com <mailto:notifications@github.com>>:

> Well, the downsides of the proposed change are that all of sudden 
> students are confronted with specifying sockets and ports on the command 
> line, and I assume that the port number may even become problematic in 
> case the mess up and run two clints simultaneously or 
> something./pelita_path/module_player.py like that.
> 

No, Pelita will manage the ports. The user will not have to change anything except for adding a little more boilerplate. As a matter of fact, the very same happens right now anyways (run a game and have a look with ps) and the Pelita code itself will not have to change except some parts can be removed.

> But, more importantly, I don't see the downsides of the current state of 
> affairs. 
> 
> >The disadvantages on Pelita’s side are:
> >a) We need to figure out the location of 
> >`./pelita_path/module_player.py`. This is harder than it seems, as it 
> >will also need to work from within the tests.
> 
> I don't see where the problem lies in this. We are doing it since the 
> very beginning of the history of Pelita, so what exactly has become 
> harder recently?

The problem is basically when Pelita is used as a library. When it is run through `./pelitagame` then finding `module_player.py` is not an issue as it should be in the same directory.
Similarly, when we run the tournament, we can also agree that `module_player.py` can be found relative to the tournament script.

However: when we run tests for the tournament we either have to use an environment variable (`PELITA_PATH`) or implement some fancy hacks because this only uses the Pelita module and therefore does not know the location of either `./pelitagame` or `module_player.py`.

This is less of an issue, of course, when it is run from git and we could obviously also give a wrapper script for running the tournament tests.

I’ll showcase the issue later with my tournament test branch.

> 
> >b) We need to provide code to specify the Python version, if this 
> >differs from Pelita’s version of Python.
> I don't see this as a problem. I would even go as far as saying that we 
> could finally drop py2 compatibility completely and move to a py3 only 
> source code. Clients will be run with the same Python version as the 
> server, full stop. So I don't see any need to specify Python versions…
> 

Dropping Python 2 would be fine with me.

**Comment by otizonaizit** [comment]
>No, Pelita will manage the ports. The user will not have to change 
>anything except for adding a little more boilerplate. As a matter of 
>fact, the very same happens right now anyways (run a game and have 
>a look with ps) and the Pelita code itself will not have to change 
>except some parts can be removed.

Ok, so I must have misinterpreted your first proposal. The only visible 
difference from the students' perspective is the little boilerplate in 
`my_player.py`? And they are still going to run games solely through 
`pelitagame`? Then I am much less critical to the proposal. The 
boilerplate code will be visible in the template they will get, so it 
shouldn't be too much of a problem.

>However: when we run tests for the tournament we either have to use an 
>environment variable (`PELITA_PATH`) or implement some fancy hacks 
>because this only uses the Pelita module and therefore does not know 
>the location of either `./pelitagame` or `module_player.py`.
Well, but thinking about it, maybe setting a PELITAPATH variable isn't 
so stupid after all… It would make many other things a bit simpler, also 
in the tournament, like storing the games, restarting the tournament, 
etc. There is also the big issue of pip-installability, of course, which 
we should address at some point. In that case PELITAPATH would be more 
problematic, right?

>>>b) We need to provide code to specify the Python version, if this 
>>>differs from Pelita’s version of Python. 
>>I don't see this as a problem. I would even go as far as saying that 
>>we could finally drop py2 compatibility completely and move to a py3 
>>only source code. Clients will be run with the same Python version as 
>>the server, full stop. So I don't see any need to specify Python 
>>versions… 
>>
>
>Dropping Python 2 would be fine with me.
OK, so maybe open a issue with that and ping the mailing list. If no one 
complains, we can set it as a milestone for next school…

**Comment by Debilski** [comment]
> Am 22.02.2016 um 10:30 schrieb Tiziano Zito <notifications@github.com>:
> 
> >No, Pelita will manage the ports. The user will not have to change 
> >anything except for adding a little more boilerplate. As a matter of 
> >fact, the very same happens right now anyways (run a game and have 
> >a look with ps) and the Pelita code itself will not have to change 
> >except some parts can be removed.
> 
> Ok, so I must have misinterpreted your first proposal. The only visible 
> difference from the students' perspective is the little boilerplate in 
> `my_player.py`? And they are still going to run games solely through 
> `pelitagame`? Then I am much less critical to the proposal. The 
> boilerplate code will be visible in the template they will get, so it 
> shouldn't be too much of a problem.

Yes, nothing else will change. One could however, think of other niceties to add to the command line such as integrated testing:

    $ ./my_player.py --test

This alone is probably not worth the change, though.

> >However: when we run tests for the tournament we either have to use an 
> >environment variable (`PELITA_PATH`) or implement some fancy hacks 
> >because this only uses the Pelita module and therefore does not know 
> >the location of either `./pelitagame` or `module_player.py`.
> Well, but thinking about it, maybe setting a PELITAPATH variable isn't 
> so stupid after all… It would make many other things a bit simpler, also 
> in the tournament, like storing the games, restarting the tournament, 
> etc. There is also the big issue of pip-installability, of course, which 
> we should address at some point. In that case PELITAPATH would be more 
> problematic, right?

Yes, the PELITAPATH (or PELITA_PATH) variable might be needed anyway because the tk graphics are also started via an external script. How do other projects do that when they need to rely on several scripts/binaries?

Also, to the whole proposal: I am undecided myself which is the better way to do it. I like it that right now we can specify a *module* on the command line and don’t have to deal with mundane things such as filenames. Typing

     # assuming ../path_to/MyPlayer/ contains a __init__.py
    $ ./pelitagame ../path_to/MyPlayer

vs

    $ ./pelitagame ../path_to/MyPlayer/player.py

feels much nicer.

The other problem: What is the right way to do in this situation:

    $ head -1 pelitagame ../path_to/MyPlayer/player.py
    #!/usr/bin/env python3
    #!/usr/bin/env python3

    $ anaconda-python3 pelitagame ../path_to/MyPlayer/player.py


> >Dropping Python 2 would be fine with me.
> OK, so maybe open a issue with that and ping the mailing list. If no one 
> complains, we can set it as a milestone for next school…

It is already there, of course :)

https://github.com/ASPP/pelita/issues/323

**Comment by Debilski** [comment]
Revisit (and most likely put on hold) after #388 is merged.

**Comment by otizonaizit** [comment]
This is a complication that has no reason to be introduced. The additional flexibility that this proposal introduces is, during ASPP, not advisable, outside of ASPP of a completely theoretical nature.
